### PR TITLE
executor: fix bug for virtual column in transaction

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve? 
When reading a virtual column in non-readonly transaction, TiDB didn't compute the virtual column so that we will get `nil` insteal of real data. 
A simple reproduce like this:

```
create table t (id int, id_gen int as (id+1), unique (id_gen));
begin;
insert into t(id) values (1);
select * from t; 
```

### What is changed and how it works?
During calling Next() in unionScanExec, compute virtual columns then do selection if exist.

### Check List 

Tests 

 - Integration test

Code changes

Side effects

Related changes

Release note

### CODE REVIEW

**Comment by wjhuang2016** [comment]
//run-all-tests

File: executor/union_scan.go [hunk 1]
```diff
@@ -120,6 +123,13 @@ func (us *UnionScanExec) Open(ctx context.Context) error {
 func (us *UnionScanExec) open(ctx context.Context) error {
 	var err error
 	reader := us.children[0]
+
+	// If the push-downed condition contains virtual column, we may build a selection upon reader. Since unionScanExec
+	// has already contained condition, we can ignore the selection.
+	if sel, ok := reader.(*SelectionExec); ok {
+		reader = sel.children[0]
```
Comment (tiancaiamao) [inline executor/union_scan.go]: The conditions are the same?

Comment (tiancaiamao) [reply executor/union_scan.go]: Why this is not done in the plan phase?

File: executor/union_scan_test.go [hunk 2]
```diff
@@ -92,6 +92,54 @@ func (s *testSuite7) TestDirtyTransaction(c *C) {
 	tk.MustQuery("select * from t where c1 < 5").Check(testkit.Rows("1 1"))
 	tk.MustQuery("select c2 from t").Check(testkit.Rows("1"))
 	tk.MustExec("commit")
+
+	// Test general virtual column
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (a int, b int as (a+1), c int as (b+1), index(c));")
+	tk.MustExec("begin;")
```
Comment (tiancaiamao) [inline executor/union_scan_test.go]: And also check `begin pessimistic`?

Comment (wjhuang2016) [reply executor/union_scan_test.go]: Ok, I wil add tests for it

File: executor/mem_reader.go [hunk 3]
```diff
@@ -82,18 +81,12 @@ func (m *memIndexReader) getMemRows() ([][]types.Datum, error) {
...
 
-	mutableRow := chunk.MutRowFromTypes(m.retFieldTypes)
 	err := iterTxnMemBuffer(m.ctx, m.kvRanges, func(key, value []byte) error {
 		data, err := m.decodeIndexKeyValue(key, value, tps)
 		if err != nil {
 			return err
 		}
 
-		mutableRow.SetDatums(data...)
-		matched, _, err := expression.EvalBool(m.ctx, m.conditions, mutableRow.ToRow())
```
Comment (crazycs520) [inline executor/mem_reader.go]: Move those condition check to `UnionScanExec.Next` may have a performance regression? 
Why not `EvalVirtualColumn` in `UnionScanExec.open` for memReader?

**Comment by Deardrops** [comment]
Closed by #15011

**Pull request closed without merging**